src/tools/piv-tool.c

```diff
@@ -503,29 +508,25 @@ static int gen_key(const char * key_info)
 		/* PIV returns 04||x||y  and x and y are the same size */
 		i = (int)(keydata.ecpoint_len - 1) / 2;
 		x = BN_bin2bn(keydata.ecpoint + 1, i, NULL);
-		y = BN_bin2bn(keydata.ecpoint + 1 + i, i, NULL) ;
+		y = BN_bin2bn(keydata.ecpoint + 1 + i, i, NULL);
+		free(keydata.ecpoint);
+		keydata.ecpoint_len = 0;
 		if (!x || !y) {
 			sc_log_openssl(ctx);
-			free(keydata.ecpoint);
-			keydata.ecpoint_len = 0;
 			BN_free(x);
 			BN_free(y);
-			EVP_PKEY_free(evpkey);
 			EC_GROUP_free(ecgroup);
 			EC_POINT_free(ecpoint);
 			return -1;
 		}
 		r = EC_POINT_set_affine_coordinates(ecgroup, ecpoint, x, y, NULL);
 
-		free(keydata.ecpoint);
-		keydata.ecpoint_len = 0;
 		BN_free(x);
 		BN_free(y);
 
 		if (r == 0) {
 			sc_log_openssl(ctx);
 			fprintf(stderr, "EC_POINT_set_affine_coordinates_GFp failed\n");
-			EVP_PKEY_free(evpkey);
```

Comment:
I think this free is still needed as the evpkey is allocated with the openssl < 3 unconditionally. Same above.